File: src/main/java/com/matag/game/event/EventSender.java
package com.matag.game.event;

import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Component;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.matag.game.player.Player;

import lombok.AllArgsConstructor;

/**
 * Sends events to the browser.
 */
@Component
@AllArgsConstructor
public class EventSender {

  private static final Logger LOGGER = LoggerFactory.getLogger(EventSender.class);

  private final SimpMessagingTemplate webSocketTemplate;
  private final ObjectMapper objectMapper;

  public void sendToUser(String sessionId, String username, Event event) {
    var eventString = serializeToString(event);
    if (!event.getType().equals("HEALTHCHECK")) {
      if (username != null) {
        LOGGER.info("Sending event to {} - {}: {}", sessionId, username, eventString);
      } else {
        LOGGER.info("Sending event to {}: {}", sessionId, eventString);
      }
    }
    webSocketTemplate.convertAndSendToUser(sessionId, "/events", eventString);
  }

  public void sendToUser(String sessionId, Event event) {
    sendToUser(sessionId, null, event);
  }

  public void sendToPlayer(Player player, Event event) {
    sendToUser(player.getToken().getSessionId(), player.getName(), event);
  }

  public void sendToPlayers(List<Player> players, Event event) {
    players.forEach(player -> sendToPlayer(player, event));
  }

  private String serializeToString(Object event) {
    try {
      return objectMapper.writeValueAsString(event);
    } catch (JsonProcessingException e) {
      throw new RuntimeException(e);
    }
  }
}

File: src/main/kotlin/com/matag/game/event/EventSender.kt
package com.matag.game.event

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.matag.game.player.Player
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Component
import java.util.function.Consumer

/**
 * Sends events to the browser.
 */
@Component
class EventSender {
    private val webSocketTemplate: SimpMessagingTemplate? = null
    private val objectMapper: ObjectMapper? = null

    fun sendToUser(sessionId: String, username: String?, event: Event) {
        val eventString = serializeToString(event)
        if (event.type != "HEALTHCHECK") {
            if (username != null) {
                LOGGER.info("Sending event to {} - {}: {}", sessionId, username, eventString)
            } else {
                LOGGER.info("Sending event to {}: {}", sessionId, eventString)
            }
        }
        webSocketTemplate!!.convertAndSendToUser(sessionId, "/events", eventString)
    }

    fun sendToUser(sessionId: String, event: Event) {
        sendToUser(sessionId, null, event)
    }

    fun sendToPlayer(player: Player, event: Event) {
        sendToUser(player.token.sessionId, player.name, event)
    }

    fun sendToPlayers(players: MutableList<Player?>, event: Event) {
        players.forEach(Consumer { player: Player? -> sendToPlayer(player!!, event) })
    }

    private fun serializeToString(event: Any?): String {
        try {
            return objectMapper!!.writeValueAsString(event)
        } catch (e: JsonProcessingException) {
            throw RuntimeException(e)
        }
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(EventSender::class.java)
    }
}
